server/user_db.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes"
+
+	pb "github.com/googleapis/gapic-showcase/server/genproto"
+	"google.golang.org/genproto/protobuf/field_mask"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func NewUserDb() UserDb {
+	return &userDb{
+		uid:   &uniqID{},
+		token: NewTokenGenerator(),
+		mu:    sync.Mutex{},
+		keys:  map[string]dbIndex{},
+		users: []userEntry{},
+	}
+}
+
+type UserDb interface {
+	Create(u *pb.User) (*pb.User, error)
+	Get(name string) (*pb.User, error)
+	Update(u *pb.User, f *field_mask.FieldMask) (*pb.User, error)
+	Delete(name string) error
+	List(page_size int32, token string) (*pb.ListUsersResponse, error)
+}
+
+type userEntry struct {
+	user    *pb.User
+	deleted bool
+}
+
+type dbIndex int
+
+type userDb struct {
+	uid   *uniqID
+	token TokenGenerator
+	mu    sync.Mutex
+	keys  map[string]dbIndex
+	users []userEntry
+}
+
+func (db *userDb) Create(u *pb.User) (*pb.User, error) {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	err := db.validate(u)
+	if err != nil {
+		return nil, err
+	}
+
+	// Validate Unique Fields.
+	hasSameName := func(x *pb.User) bool {
+		return u.GetDisplayName() == x.GetDisplayName()
+	}
+	if db.anyUser(hasSameName) {
+		return nil, status.Errorf(
+			codes.AlreadyExists,
+			"A user with display_name %s already exists.",
+			u.GetDisplayName())
+	}
+
+	hasSameEmail := func(x *pb.User) bool { return u.GetEmail() == x.GetEmail() }
+	if db.anyUser(hasSameEmail) {
+		return nil, status.Errorf(
+			codes.AlreadyExists,
+			"A user with the email %s already exists.",
+			u.GetEmail())
+	}
+
+	// Assign info.
+	id := db.uid.id()
+	name := fmt.Sprintf("users/%d", id)
+	now := ptypes.TimestampNow()
+
+	u.Name = name
+	u.CreateTime = now
+	u.UpdateTime = now
+
+	// Insert.
+	index := dbIndex(len(db.users))
+	db.users = append(db.users, userEntry{user: u, deleted: false})
+	db.keys[name] = index
+
+	return u, nil
+}
+
+func (db *userDb) Get(s string) (*pb.User, error) {
+	if v, ok := db.keys[s]; ok {
```

Comment:
We need mutex lock here right?

Reply:
Good Catch! I didn't need one before when I was using just a map, but when I switch to using a pseudo ordered map, I do. Fixed!